pypeman/nodes.py

```diff
@@ -899,6 +899,67 @@ def generator(msg):
         return generator(msg)
 
 
+class MsgFuncNode(BaseNode):
+    """
+    Make a node from a simple function/coroutine that takes the message
+    and modifies it.
+
+    The given function can be async or not.
+    The name for the node will automatically be using the function's name.
+
+        def do_something(msg):
+            msg.meta ...
+
+        MsgFuncNode(do_something)
+
+    See also `FuncNode` to work purely with the payload.
+    """
+
+    def __init__(self, fn, *, name: str = "", **kwargs):
+        if not name and fn.__name__ != "<lambda>":
+            name = self.__class__.__name__ + "_" + fn.__name__
+        super().__init__(name=name, **kwargs)
+        self.fn = fn
+
+    async def process(self, msg):
+        if asyncio.iscoroutinefunction(self.fn):
+            await self.fn(msg)
+        else:
+            self.fn(msg)
+        return msg
+
+
+class FuncNode(BaseNode):
+    """
+    Make a node from a simple function/coroutine that takes the message's
+    payload and returns a (potentially identical) payload.
+
+    The given function can be async or not.
+    The name for the node will automatically be using the function's name.
+
+        def do_something(payload):
+            return payload + 1
+
+        FuncNode(do_something)
+
+    See also `MsgFuncNode` to get the whole message.
+    """
+
+    def __init__(self, fn, *, name: str = "", **kwargs):
+        if not name and fn.__name__ != "<lambda>":
+            name = self.__class__.__name__ + "_" + fn.__name__
+        super().__init__(name=name, **kwargs)
+        self.fn = fn
+
+    async def process(self, msg):
+        msg.payload = (
+            await self.fn(msg.payload)
+            if asyncio.iscoroutinefunction(self.fn)
```

Comment:
can close.

Interesting! I would not have expected this works. but great if it does. I learnt something.


I'd have thought that

`v = await f() if conditition else f()`

is equivalent to

`v = await ( f() if condition else f() )`

and not

`v = (await f()) if condition else f()`

Reply:
yeah `await` is actually [one of the highest precedence](https://docs.python.org/3/reference/expressions.html#operator-precedence), and `if-else` one of the lowest; this link is a good ref to have/know about